File: examples/api_client.py
#!/usr/bin/env python3
"""
RuleGems HTTP API 客户端示例

这是一个完整的 Python 客户端示例，展示了如何调用 RuleGems 插件的 HTTP API。
"""

import requests
import json
import sys
from typing import Optional, Dict, Any


class RuleGemsAPIClient:
    """RuleGems HTTP API 客户端"""
    
    def __init__(self, base_url: str, token: Optional[str] = None):
        """
        初始化 API 客户端
        
        Args:
            base_url: API 基础地址，例如 "http://localhost:8080"
            token: 访问令牌（可选）
        """
        self.base_url = base_url.rstrip('/')
        self.token = token
        self.session = requests.Session()
        
        # 设置通用请求头
        self.session.headers.update({
            'User-Agent': 'RuleGemsAPIClient/1.0.0',
            'Accept': 'application/json'
        })
    
    def _build_params(self, **kwargs) -> Dict[str, str]:
        """构建请求参数"""
        params = {k: v for k, v in kwargs.items() if v is not None}
        if self.token:
            params['token'] = self.token
        return params
    
    def get_player_gems(self, player_name: Optional[str] = None, 
                       player_uuid: Optional[str] = None) -> Dict[str, Any]:
        """
        获取玩家宝石数据
        
        Args:
            player_name: 玩家名称
            player_uuid: 玩家 UUID
            
        Returns:
            API 响应数据
            
        Raises:
            requests.RequestException: 网络请求错误
        """
        if not player_name and not player_uuid:
            raise ValueError("必须提供 player_name 或 player_uuid 参数")
        
        params = self._build_params(player=player_name, uuid=player_uuid)
        
        try:
            response = self.session.get(f"{self.base_url}/api/gems", params=params)
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            print(f"获取玩家宝石数据失败: {e}")
            raise
    
    def get_online_players(self) -> Dict[str, Any]:
        """
        获取在线玩家列表
        
        Returns:
            API 响应数据
        """
        params = self._build_params()
        
        try:
            response = self.session.get(f"{self.base_url}/api/players", params=params)
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            print(f"获取在线玩家列表失败: {e}")
            raise
    
    def get_server_status(self) -> Dict[str, Any]:
        """
        获取服务器状态
        
        Returns:
            API 响应数据
        """
        params = self._build_params()
        
        try:
            response = self.session.get(f"{self.base_url}/api/status", params=params)
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            print(f"获取服务器状态失败: {e}")
            raise
    
    def test_connection(self) -> bool:
        """
        测试 API 连接
        
        Returns:
            连接是否成功
        """
        try:
            response = self.get_server_status()
            return response.get('success', False)
        except:
            return False


def print_json(data: Dict[str, Any], title: str = ""):
    """格式化打印 JSON 数据"""
    if title:
        print(f"\n{'='*50}")
        print(f"{title}")
        print(f"{'='*50}")
    print(json.dumps(data, indent=2, ensure_ascii=False))


def main():
    """主函数 - 演示 API 客户端使用"""
    
    # 配置参数
    BASE_URL = "http://localhost:8080"
    TOKEN = "your_secret_token"  # 替换为实际的访问令牌
    
    # 创建 API 客户端
    api = RuleGemsAPIClient(BASE_URL, TOKEN)
    
    print("RuleGems API 客户端演示")
    print("=" * 50)
    
    # 测试连接
    print("1. 测试 API 连接...")
    if api.test_connection():
        print("✅ API 连接成功")
    else:
        print("❌ API 连接失败，请检查配置")
        sys.exit(1)
    
    # 获取服务器状态
    print("\n2. 获取服务器状态...")
    try:
        status = api.get_server_status()
        if status['success']:
            print_json(status['data'], "服务器状态")
        else:
            print(f"❌ 获取服务器状态失败: {status['message']}")
    except Exception as e:
        print(f"❌ 获取服务器状态异常: {e}")
    
    # 获取在线玩家列表
    print("\n3. 获取在线玩家列表...")
    try:
        players = api.get_online_players()
        if players['success']:
            print_json(players['data'], "在线玩家列表")
            
            # 如果有在线玩家，获取第一个玩家的宝石数据
            if players['data'].get('players'):
                first_player = players['data']['players'][0]
                player_name = first_player.get('playerName')
                player_uuid = first_player.get('playerId')
                
                print(f"\n4. 获取玩家 '{player_name}' 的宝石数据...")
                try:
                    gems = api.get_player_gems(player_name=player_name)
                    if gems['success']:
                        print_json(gems['data'], f"玩家 {player_name} 的宝石数据")
                    else:
                        print(f"❌ 获取玩家宝石数据失败: {gems['message']}")
                except Exception as e:
                    print(f"❌ 获取玩家宝石数据异常: {e}")
        else:
            print(f"❌ 获取在线玩家列表失败: {players['message']}")
    except Exception as e:
        print(f"❌ 获取在线玩家列表异常: {e}")
    
    # 演示错误处理
    print("\n5. 演示错误处理...")
    
    # 测试不存在的玩家
    try:
        invalid_player = api.get_player_gems(player_name="NonExistentPlayer")
        if not invalid_player['success']:
            print(f"✅ 预期错误处理: {invalid_player['message']}")
    except Exception as e:
        print(f"❌ 错误处理异常: {e}")
    
    print("\n" + "=" * 50)
    print("演示完成")


if __name__ == "__main__":
    # 检查依赖
    try:
        import requests
    except ImportError:
        print("❌ 缺少依赖: requests")
        print("请安装: pip install requests")
        sys.exit(1)
    
    main()

File: examples/api_client.js
/**
 * RuleGems HTTP API 客户端示例 (JavaScript)
 * 
 * 这是一个完整的 JavaScript 客户端示例，展示了如何调用 RuleGems 插件的 HTTP API。
 * 适用于浏览器环境和 Node.js 环境。
 */

class RuleGemsAPIClient {
    /**
     * 初始化 API 客户端
     * @param {string} baseUrl - API 基础地址，例如 "http://localhost:8080"
     * @param {string} [token] - 访问令牌（可选）
     */
    constructor(baseUrl, token = null) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.token = token;
        this.headers = {
            'User-Agent': 'RuleGemsAPIClient/1.0.0',
            'Accept': 'application/json'
        };
    }

    /**
     * 构建请求参数
     * @param {Object} params - 参数对象
     * @returns {URLSearchParams}
     */
    _buildParams(params = {}) {
        const searchParams = new URLSearchParams();
        
        // 添加参数
        Object.entries(params).forEach(([key, value]) => {
            if (value !== null && value !== undefined) {
                searchParams.append(key, value.toString());
            }
        });
        
        // 添加令牌
        if (this.token) {
            searchParams.append('token', this.token);
        }
        
        return searchParams;
    }

    /**
     * 发送 HTTP 请求
     * @param {string} url - 请求 URL
     * @param {Object} [params] - 请求参数
     * @returns {Promise<Object>}
     */
    async _request(url, params = {}) {
        const queryString = this._buildParams(params).toString();
        const fullUrl = queryString ? `${url}?${queryString}` : url;
        
        try {
            const response = await fetch(fullUrl, {
                method: 'GET',
                headers: this.headers
            });
            
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }
            
            return await response.json();
        } catch (error) {
            console.error(`API 请求失败: ${error.message}`);
            throw error;
        }
    }

    /**
     * 获取玩家宝石数据
     * @param {string} [playerName] - 玩家名称
     * @param {string} [playerUuid] - 玩家 UUID
     * @returns {Promise<Object>}
     */
    async getPlayerGems(playerName = null, playerUuid = null) {
        if (!playerName && !playerUuid) {
            throw new Error('必须提供 playerName 或 playerUuid 参数');
        }
        
        const params = {
            player: playerName,
            uuid: playerUuid
        };
        
        return await this._request(`${this.baseUrl}/api/gems`, params);
    }

    /**
     * 获取在线玩家列表
     * @returns {Promise<Object>}
     */
    async getOnlinePlayers() {
        return await this._request(`${this.baseUrl}/api/players`);
    }

    /**
     * 获取服务器状态
     * @returns {Promise<Object>}
     */
    async getServerStatus() {
        return await this._request(`${this.baseUrl}/api/status`);
    }

    /**
     * 测试 API 连接
     * @returns {Promise<boolean>}
     */
    async testConnection() {
        try {
            const status = await this.getServerStatus();
            return status.success === true;
        } catch {
            return false;
        }
    }
}

/**
 * 格式化打印 JSON 数据
 * @param {Object} data - 要打印的数据
 * @param {string} title - 标题
 */
function printJson(data, title = '') {
    if (title) {
        console.log('='.repeat(50));
        console.log(title);
        console.log('='.repeat(50));
    }
    console.log(JSON.stringify(data, null, 2));
}

/**
 * 主函数 - 演示 API 客户端使用
 */
async function main() {
    // 配置参数
    const BASE_URL = 'http://localhost:8080';
    const TOKEN = 'your_secret_token'; // 替换为实际的访问令牌
    
    // 创建 API 客户端
    const api = new RuleGemsAPIClient(BASE_URL, TOKEN);
    
    console.log('RuleGems API 客户端演示 (JavaScript)');
    console.log('='.repeat(50));
    
    // 测试连接
    console.log('1. 测试 API 连接...');
    const isConnected = await api.testConnection();
    if (isConnected) {
        console.log('✅ API 连接成功');
    } else {
        console.log('❌ API 连接失败，请检查配置');
        return;
    }
    
    // 获取服务器状态
    console.log('\n2. 获取服务器状态...');
    try {
        const status = await api.getServerStatus();
        if (status.success) {
            printJson(status.data, '服务器状态');
        } else {
            console.log(`❌ 获取服务器状态失败: ${status.message}`);
        }
    } catch (error) {
        console.log(`❌ 获取服务器状态异常: ${error.message}`);
    }
    
    // 获取在线玩家列表
    console.log('\n3. 获取在线玩家列表...');
    try {
        const players = await api.getOnlinePlayers();
        if (players.success) {
            printJson(players.data, '在线玩家列表');
            
            // 如果有在线玩家，获取第一个玩家的宝石数据
            if (players.data.players && players.data.players.length > 0) {
                const firstPlayer = players.data.players[0];
                const playerName = firstPlayer.playerName;
                const playerUuid = firstPlayer.playerId;
                
                console.log(`\n4. 获取玩家 '${playerName}' 的宝石数据...`);
                try {
                    const gems = await api.getPlayerGems(playerName);
                    if (gems.success) {
                        printJson(gems.data, `玩家 ${playerName} 的宝石数据`);
                    } else {
                        console.log(`❌ 获取玩家宝石数据失败: ${gems.message}`);
                    }
                } catch (error) {
                    console.log(`❌ 获取玩家宝石数据异常: ${error.message}`);
                }
            }
        } else {
            console.log(`❌ 获取在线玩家列表失败: ${players.message}`);
        }
    } catch (error) {
        console.log(`❌ 获取在线玩家列表异常: ${error.message}`);
    }
    
    // 演示错误处理
    console.log('\n5. 演示错误处理...');
    
    // 测试不存在的玩家
    try {
        const invalidPlayer = await api.getPlayerGems('NonExistentPlayer');
        if (!invalidPlayer.success) {
            console.log(`✅ 预期错误处理: ${invalidPlayer.message}`);
        }
    } catch (error) {
        console.log(`❌ 错误处理异常: ${error.message}`);
    }
    
    console.log('\n' + '='.repeat(50));
    console.log('演示完成');
}

// 浏览器环境使用示例
if (typeof window !== 'undefined') {
    // 在浏览器中运行时，将函数暴露给全局作用域
    window.RuleGemsAPIClient = RuleGemsAPIClient;
    window.demoRuleGemsAPI = main;
    
    console.log('RuleGems API 客户端已加载到浏览器环境');
    console.log('使用方法: await demoRuleGemsAPI()');
}

// Node.js 环境使用示例
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        RuleGemsAPIClient,
        main
    };
    
    // 如果直接运行此文件，则执行演示
    if (require.main === module) {
        main().catch(error => {
            console.error('演示失败:', error);
            process.exit(1);
        });
    }
}

// 浏览器环境自动运行演示（可选）
if (typeof window !== 'undefined' && window.location.search.includes('demo')) {
    console.log('自动运行演示...');
    main().catch(error => console.error('演示失败:', error));
}